Realm/RLMAccessor.mm

```diff
@@ -692,3 +562,295 @@ id RLMDynamicGetByName(__unsafe_unretained RLMObjectBase *const obj, __unsafe_un
+}
+template<>
+long long RLMAccessorContext::unbox(__unsafe_unretained id const v, bool, bool) {
+    checkType([v isKindOfClass:[NSNumber class]], v, @"int");
+    return [v longLongValue];
+}
+template<>
+realm::BinaryData RLMAccessorContext::unbox(id v, bool, bool) {
+    v = RLMCoerceToNil(v);
+    checkType(!v || [v respondsToSelector:@selector(bytes)], v, @"data");
+    return RLMBinaryDataForNSData(v);
+}
+template<>
+realm::StringData RLMAccessorContext::unbox(id v, bool, bool) {
+    v = RLMCoerceToNil(v);
+    checkType(!v || [v respondsToSelector:@selector(UTF8String)], v, @"string");
+    return RLMStringDataWithNSString(v);
+}
+
+template<typename Fn>
+static auto to_optional(__unsafe_unretained id const value, Fn&& fn) {
+    id v = RLMCoerceToNil(value);
+    return v && v != NSNull.null ? realm::util::make_optional(fn(v)) : realm::util::none;
+}
+
+template<>
+realm::util::Optional<bool> RLMAccessorContext::unbox(__unsafe_unretained id const v, bool, bool) {
+    return to_optional(v, [&](__unsafe_unretained id v) {
+        checkType([v respondsToSelector:@selector(boolValue)], v, @"bool?");
+        return (bool)[v boolValue];
+    });
+}
+template<>
+realm::util::Optional<double> RLMAccessorContext::unbox(__unsafe_unretained id const v, bool, bool) {
+    return to_optional(v, [&](__unsafe_unretained id v) {
+        checkType([v respondsToSelector:@selector(doubleValue)], v, @"double?");
+        return [v doubleValue];
+    });
+}
+template<>
+realm::util::Optional<float> RLMAccessorContext::unbox(__unsafe_unretained id const v, bool, bool) {
+    return to_optional(v, [&](__unsafe_unretained id v) {
+        checkType([v respondsToSelector:@selector(floatValue)], v, @"float?");
+        return [v floatValue];
+    });
+}
+template<>
+realm::util::Optional<int64_t> RLMAccessorContext::unbox(__unsafe_unretained id const v, bool, bool) {
+    return to_optional(v, [&](__unsafe_unretained id v) {
+        checkType([v respondsToSelector:@selector(longLongValue)], v, @"int?");
+        return [v longLongValue];
+    });
+}
+
+template<>
+realm::RowExpr RLMAccessorContext::unbox(__unsafe_unretained id const v, bool create, bool update) {
+    RLMObjectBase *link = RLMDynamicCast<RLMObjectBase>(v);
+    if (!link) {
+        if (!create)
+            return realm::RowExpr();
+        return RLMCreateObjectInRealmWithValue(_realm, _info.rlmObjectSchema.className, v, update)->_row;
+    }
+
+    if (link.isInvalidated) {
+        if (create) {
+            @throw RLMException(@"Adding a deleted or invalidated object to a Realm is not permitted");
+        }
+        else {
+            @throw RLMException(@"Object has been invalidated");
+        }
+    }
+
+    if (![link->_objectSchema.className isEqualToString:_info.rlmObjectSchema.className]) {
+        if (create && !_promote_existing)
+            return RLMCreateObjectInRealmWithValue(_realm, _info.rlmObjectSchema.className, link, update)->_row;
+        return link->_row;
```

Comment:
The code doing things with the row validates that the row belongs to the correct table.